### PR TITLE
Document available services and their routes

### PR DESCRIPTION
**Pull request by delvh**

### CODE REVIEW

File: docs/dev-manuals/architecture/services.md [hunk 1]
```diff
@@ -0,0 +1,76 @@
+# Services
+
```
Comment (lesser1) [inline docs/dev-manuals/architecture/services.md:2]: Why is this a numerical list but every item of it starts at 1.?

Comment (nk-coding) [reply docs/dev-manuals/architecture/services.md:2]: In the rendered version, markdown automatically increases the number. This makes maintaining the source easier

File: docs/dev-manuals/architecture/services.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
+# Services
```
Comment (nk-coding) [inline docs/dev-manuals/architecture/services.md:1]: where is the position meta data stuff?

File: docs/dev-manuals/architecture/services.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+1. Product
+   - stores concrete product information - name, description, image locations, and product variants, i.e. XXL T-Shirt in red - for each product
+   - does not store individual instances but the overall descriptions of an item. For that see `inventory`
+   - basepath `/products`
+   - CRUD on `/products`, CRUD on `/products/<product>/variants`
```
Comment (legendofa) [inline docs/dev-manuals/architecture/services.md:7]: How to query different product variant versions?

File: docs/dev-manuals/architecture/services.md [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+1. Address
+   - Stores possible addresses, and to which user they belong
+   - routes `/user/<username>/addresses`
+1. Media
+   - file/object storage/db
+   - seller can upload images and retrieve URL of / delete the uploaded image
+   - basepath `/media`
+   - routes: CRUD on `/media/<file>`
+1. Discount
+   - Stores available discounts per user and/or product
```
Comment (legendofa) [inline docs/dev-manuals/architecture/services.md:31]: Must store both `Coupon` and `Discounts` and enable CRUD

File: docs/dev-manuals/architecture/services.md [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+   - basepath: `/wishlist`
+   - routes: CRUD on `/user/<username>/wishlist/items`
+1. Review
+   - associates a user review with a product
+   - provides summary data
+   - routes:
+     - CRUD on `/product/<product>/reviews` - _list_ return all reviews on this product, no matter the variant. _Create_ needs the precise product variant too.
+     - `/variant/<product variant>/review` - all reviews for this specific variant
+1. Category
+   - stores the available categories of items
```
Comment (legendofa) [inline docs/dev-manuals/architecture/services.md:52]:
```suggestion
   - stores the available categories of products
```

File: docs/dev-manuals/architecture/services.md [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+   - file/object storage/db
+   - seller can upload images and retrieve URL of / delete the uploaded image
+   - basepath `/media`
+   - routes: CRUD on `/media/<file>`
+1. Discount
+   - Stores available discounts per user and/or product
+   - basepath: `/discounts`
+   - routes: CRUD on `/discounts`, `/user/<username>/order-discounts`
+1. Cart
+   - stores items that are in the pending order (the next one to be bought)
```
Comment (legendofa) [inline docs/dev-manuals/architecture/services.md:35]: Why pending order? I think the order only matters to display which items were added first

File: docs/dev-manuals/architecture/services.md [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+1. Product
+   - stores concrete product information - name, description, image locations, and product variants, i.e. XXL T-Shirt in red - for each product
+   - does not store individual instances but the overall descriptions of an item. For that see `inventory`
+   - basepath `/products`
+   - CRUD on `/products`, CRUD on `/products/<product>/variants`
+1. Account
+   - Handles session information for a user, including login/registration
+   - In our case, most likely a Keycloak
+   - basepath `/account`
+   - routes `/account/register`, `/account/login`, \<TODO: Path to query JWT token of a user\>
```
Comment (legendofa) [inline docs/dev-manuals/architecture/services.md:12]: How about adding `/accounts` to query all accounts for admin purposes?

File: docs/dev-manuals/architecture/services.md [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+1. User
+   - user settings (whenJoined, birthday, gender)
+   - CRUD on `/user/<username>/settings`
+1. Notification
+   - Provides notfications to show a user
+   - mail replacement
+   - supposed to cover everything found under a typical bell icon
+   - basepath `/notifications`
+   - routes `/user/<username>/notifications?state=<state>`, `/user/<username>/notifications/<notification>/change-read`
```
Comment (legendofa) [inline docs/dev-manuals/architecture/services.md:21]:
```suggestion
   - routes `/user/<username>/notifications?isRead=<state>`, `/user/<username>/notifications/<notification>/isRead=<state>`
```

File: docs/dev-manuals/architecture/services.md [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+1. Address
+   - Stores possible addresses, and to which user they belong
+   - routes `/user/<username>/addresses`
```
Comment (legendofa) [inline docs/dev-manuals/architecture/services.md:24]: Addresses should be queried without a reference to the user, as they are used for orders/returns, etc. I would suggest an ID:

```suggestion
   - routes `/addresses/<id>`, `/addresses`
```

File: docs/dev-manuals/architecture/services.md (hunk 1 above)
Comment (lesser1) [reply docs/dev-manuals/architecture/services.md:2]: That's awesome! I like it

File: docs/dev-manuals/architecture/services.md [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+   - basepath: `/cart`
+   - routes: CRUD on `/user/<username>/cart/items`
+1. Wishlist
+   - stores items that the user wants to buy at some later point in time
+   - wishlist is a shopping cart with two exceptions: it does not store an amount of items to buy, and items on a wishlist will not be bought with the next order
+   - any user can have multiple wishlists
+   - wishlists can potentially be used for recommendations
+   - basepath: `/wishlist`
+   - routes: CRUD on `/user/<username>/wishlist/items`
+1. Review
```
Comment (legendofa) [inline docs/dev-manuals/architecture/services.md:45]: What about querying the reviews of a specific user: `/user/<username>/reviews`

File: docs/dev-manuals/architecture/services.md [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+1. Category
+   - stores the available categories of items
+   - routes: CRUD on `/categories`
+1. Inventory
+   - stores each available product digitally
+   - handles state of item (blocked when sold but not yet delivered)
+   - routes: CRUD on `/inventory/<product variant>`, `/inventory/<product variant>/(un)block`
+1. Tax
+   - stores the tax to subtract from any product variant
+   - routes: CRUD on `/taxes`
```
Comment (legendofa) [inline docs/dev-manuals/architecture/services.md:60]:
```suggestion
  - routes: CRUD on `/tax-rates`
```

File: docs/dev-manuals/architecture/services.md [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+     - CRUD on `/product/<product>/reviews` - _list_ return all reviews on this product, no matter the variant. _Create_ needs the precise product variant too.
+     - `/variant/<product variant>/review` - all reviews for this specific variant
+1. Category
+   - stores the available categories of items
+   - routes: CRUD on `/categories`
+1. Inventory
+   - stores each available product digitally
+   - handles state of item (blocked when sold but not yet delivered)
+   - routes: CRUD on `/inventory/<product variant>`, `/inventory/<product variant>/(un)block`
+1. Tax
```
Comment (legendofa) [inline docs/dev-manuals/architecture/services.md:58]: How about querying different tax rate versions?

File: docs/dev-manuals/architecture/services.md [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+   - routes: CRUD on `/taxes`
+1. Invoice
+   - Generates an invoice for an order
+   - routes: CRUD on `/invoices/<order>`
+1. Order
+   - stores items that have been ordered simultaneously
+   - routes: CRUD on `/user/<username>/orders`
+1. Payment
+   - handles payment of any order
+   - routes: CRUD on `/payment/providers`, CRUD on `/user/<username>/payments`, `/payments/<payment>/receipt`
```
Comment (legendofa) [inline docs/dev-manuals/architecture/services.md:69]: I am not sure if we should query `providers` under `payment`. It this is the case then `payment` should also return a useful resource

**Review by legendofa — CHANGES_REQUESTED** [review]
It might be more easy to first create API endpoints for the resources that are fundamental for each service. Also each parent folder should be a plural: `review -> reviews` For example: 

Review service: `/reviews/<id>`
Payment service: `payments/<id>`, `payment-providers/<id>`

And then build user based functionality upon that, if necessary:

Review service: `users/<username>/reviews/<id>`
Payment service: `users/<username>/payments/<id>`, `payment-providers/id`

**Comment by legendofa** [comment]
I guess this is now deprecated with GraphQL Federation, shall we close it?

**Comment by delvh** [comment]
I still need to update it.
After all, such an overview is still necessary, but better.

**Pull request closed without merging**